components/python/python-312/Makefile

```diff
@@ -0,0 +1,208 @@
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright (c) 2011, 2020, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2019, Michal Nowak
+# Copyright (c) 2021, Aurelien Larcher
+#
+
+OPENSSL_VERSION=3.1
+include ../../../make-rules/shared-macros.mk
+
+COMPONENT_NAME=	        Python
+COMPONENT_VERSION=      3.12.1
+COMPONENT_SUMMARY=      The Python interpreter, libraries and utilities
+COMPONENT_PROJECT_URL=  https://python.org/
+COMPONENT_SRC=          $(COMPONENT_NAME)-$(COMPONENT_VERSION)
+COMPONENT_ARCHIVE=      $(COMPONENT_SRC).tar.xz
+COMPONENT_ARCHIVE_HASH= sha256:8dfb8f426fcd226657f9e2bd5f1e96e53264965176fa17d32658e873591aeb21
+COMPONENT_ARCHIVE_URL = $(COMPONENT_PROJECT_URL)ftp/python/$(COMPONENT_VERSION)/$(COMPONENT_ARCHIVE)
+COMPONENT_FMRI=			runtime/python-312
+COMPONENT_CLASSIFICATION= Development/Python
+COMPONENT_LICENSE=      PSFv2
+COMPONENT_LICENSE_FILE=	python312.license
+
+CONFIGURE_FIX_LIBTOOL_RPATH = no
+
+# Define to avoid modifications at each major bump
+COMPONENT_PYVER= 3.12
+
+include $(WS_MAKE_RULES)/common.mk
+
+PATH= $(PATH.gnu)
+
+# Need to preserve timestamp for Grammar files.  If the pickle files are older,
+# Python will try to rebuild them.
+PKGSEND_PUBLISH_OPTIONS += -T \*Grammar\*.txt
+PKGSEND_PUBLISH_OPTIONS += -T \*Grammar\*.pickle
+
+# We patch auto* files, so regenerate headers and configure
+COMPONENT_PREP_ACTION = (cd $(@D) ; autoheader ; autoconf -f)
+
+# we don't want to leak $(CC_BITS) into BASECFLAGS as it causes problems with
+# python-config
+CC += $(CFLAGS)
+
+gcc_OPT = -O2
+
+CPPFLAGS += -IPython
+
+# to find the ncurses headers
+CPPFLAGS += -I/usr/include/ncurses
+# enable large files how they did in JDS
+CPPFLAGS += -D_LARGEFILE64_SOURCE
+
+# libffi for _ctypes
+CPPFLAGS += $(shell pkg-config --cflags-only-I libffi)
+
+# because python links with $(CC) ... $(LDFLAGS) ...
+LDFLAGS = $(CC_BITS) $(CC_PIC)
+
+# build pic
+CFLAGS += $(CC_PIC)
+CFLAGS += $(CFLAGS.$(MACH))
+
+# XPG6
+CFLAGS += $(CPP_XPG6MODE)
+
+# The python build is profile-guided for gcc.  Use of xprofile requires that
+# the same options be used during compilation and linking.  The targets chosen
+# are based on Solaris 11 minimum supported system requirements.
+XPROFILE_DIR= $(BUILD_DIR_$(BITS))/.profile
+
+# PYTHONPATH in the environment can be harmful, but setting it to empty via
+# _INSTALL_ENV causes problems too, so just ignore the entire environment.
+# (see 20396493) Because of this, we have to explicitly specify PATH in
+# multiple places below. Also up the limit of maximum opened files for
+# testing purposes and pgo.
+ENV := ulimit -n 16384; $(ENV) -i
+
+# Configure script should look into sbin folder when looking for dtrace
+CONFIGURE_ENV += PATH="$(PATH):/usr/sbin"
+
+# Disable epoll on illumos
+CONFIGURE_ENV += ac_cv_func_epoll_ctl=no
+CONFIGURE_ENV += ac_cv_func_epoll_create1=no
+
+CONFIGURE_ENV += ac_cv_func_getentropy=no
```

Comment:
This is similar to line 119 below, but it is commented there.  The linked bug (see below) is closed as resolved (with fix merged) for many years now.  Could you please make sure this one is still needed?  If so, then please add some reasonable comment explaining the reason.

Thank you.

Reply:
As far as I can tell disabling getentropy is not needed and these lines will be gone in the next push.   Truss on a short python script shows that os.urandom() generates a call into the getrandom() syscall passing a zero flag (which is the desirable fast-path, unlikely to block) rather than calling getentropy().

Reply:
> As far as I can tell disabling getentropy is not needed and these lines will be gone in the next push. Truss on a short python script shows that os.urandom() generates a call into the getrandom() syscall passing a zero flag (which is the desirable fast-path, unlikely to block) rather than calling getentropy().

I assume you trussed with these lines removed... (just to be sure we are on the same page :-))

Reply:
The code in `Python/bootstrap_hash.c` is structured such that it will only build code to use getentropy if getrandom isn't present:

```
#if defined(HAVE_GETRANDOM) || defined(HAVE_GETRANDOM_SYSCALL)
#define PY_GETRANDOM 1
...
#elif defined(HAVE_GETENTROPY)
#define PY_GETENTROPY 1
...
#endif /* defined(HAVE_GETENTROPY) && !(defined(__sun) && defined(__SVR4)) */
```
Just to be sure I'll test this again with a build with the lines removed.

Reply:
> Just to be sure I'll test this again with a build with the lines removed.

I've retested.  As expected, removing the `...getentropy=no` lines from the Makefile had no effect on the behavior - `os.urandom()` results in a `getrandom()` syscall with them either present or  removed.;